src/compiler/types/stencil-types.ts

```diff
@@ -30,6 +30,81 @@ export const updateStencilTypesImports = (typesDir: string, dtsFilePath: string,
   return dtsContent;
 };
 
+/**
+ * Utility for ensuring that naming collisions do not appear in type declaration files for a component's class members
+ * decorated with @Prop, @Event, and @Method
+ * @param typeReferences all type names used by a component class member
+ * @param typeImportData locally/imported/globally used type names, which may be used to prevent naming collisions
+ * @param sourceFilePath the path to the source file of a component using the type being inspected
+ * @param initialType the name of the type that may be updated
+ * @returns the updated type name, which may be the same as the initial type name provided as an argument to this
+ * function
+ */
+export const updateTypeIdentifierNames = (
+  typeReferences: d.ComponentCompilerTypeReferences,
+  typeImportData: d.TypesImportData,
+  sourceFilePath: string,
+  initialType: string
+): string => {
+  let currentTypeName = initialType;
+
+  // iterate over each of the type references, as there may be >1 reference to inspect
+  for (let typeReference of Object.values(typeReferences)) {
+    const importResolvedFile = getTypeImportPath(typeReference.path, sourceFilePath);
+
+    if (!typeImportData.hasOwnProperty(importResolvedFile)) {
+      continue;
+    }
+
+    for (let typesImportDatumElement of typeImportData[importResolvedFile]) {
+      currentTypeName = updateTypeName(currentTypeName, typesImportDatumElement);
+    }
+  }
+  return currentTypeName;
```

Comment:
one thing I'm wondering - this function is going to be called for every component prop, every event, and so on, and then for each of those we're doing two nested for loops. Are we worried about the performance impact of doing that? Do you think it's worthwhile / necessary to do a little measurement to see if there's an impact? When I see a `for ... of` within a `for ... of` and then realize that this function is itself called within a loop I wonder, but I don't feel like I have enough context to know how the things we're looping through are going to be. So possibly not a concern, but wanted to flag it to see your thoughts and make sure I'm understanding what's going on.

Reply:
> Are we worried about the performance impact of doing that?

A little, yes. Specifically, I'm concerned about the cost of these two lines in `updateTypeName`:
```ts
  const typeNameRegex = new RegExp(`${typeAlias.localName}\\b${endingStrChar}`, 'g');
  return currentTypeName.replace(typeNameRegex, typeAlias.importName);
```
where we build this regex and run `replace` globally on `currentTypeName`. I tried to add as many cases where we can avoid this work as possible (if there's no alias to replace a collision with, if we don't see any collisions at all, etc.), but we still see something like `O(m * n * p)` performance where:
- `m` is the size of `ComponentCompilerTypeReferences` 
- `n` is the size of `TypesImportData` 
- `p` is the length of the string type of the collision (maybe, the perf implications may be worse here with allocating memory for a new string)

My hope is the `m` and `n` are relatively small since we evaluate this on a per component (and therefore per file) basis. `p` is a bit trickier to optimize ATM, and is going to require some rework that I planned in STENCIL-418+STENCIL-419 (we'd need to move some of this cost to allocating/storing the actual types instead of flattening them to strings).

Reply:
Ok that all makes sense, I was thinking that given that we're going through a component at a time it shouldn't be a _huge_ issue, but if stencil users have some enormous components with 30 or 40 imports we might see some real impact there.

Incidentally, this seems like exactly the type of situation where some automated performance testing against a large stencil project would be super handy!